File: java-sdk/sdk/src/main/kotlin/org/apache/airflow/sdk/execution/Serde.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.airflow.sdk.execution

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.airflow.sdk.Bundle
import org.apache.airflow.sdk.DagDef
import org.apache.airflow.sdk.TaskDef
import org.apache.airflow.sdk.execution.comm.DagFileParseRequest
import org.apache.airflow.sdk.internal.Field
import org.apache.airflow.sdk.internal.SchemaFields
import java.nio.file.InvalidPathException
import java.nio.file.Paths
import java.time.Duration
import java.time.Instant
import java.time.OffsetDateTime

// Serializes Dags to Airflow DagSerialization v3 JSON, mirroring the Go SDK's
// serde (go-sdk/pkg/execution/serde.go), which in turn matches Python's
// DagSerialization output.

// Per-Dag defaults that Python resolves from [core] config when the Dag does
// not override them. The serializer always emits these fields (they have no
// JSON-schema default to omit against), so we fall back to the same values.
private const val DEFAULT_MAX_ACTIVE_TASKS_PER_DAG = 16 // [core] max_active_tasks_per_dag
private const val DEFAULT_MAX_ACTIVE_RUNS_PER_DAG = 16 // [core] max_active_runs_per_dag

private val defaultsMapper = ObjectMapper()

/**
 * Processes a [DagFileParseRequest] by serialising every Dag registered on
 * [bundle] to DagSerialization v3 and returning the result as a
 * DagFileParsingResult body.
 */
internal fun parseDags(
  bundle: Bundle,
  request: DagFileParseRequest,
): Map<String, Any?> {
  val fileloc = request.file ?: ""
  val relativeFileloc = computeRelativeFileloc(fileloc, request.bundlePath)
  val serializedDags =
    bundle.dags.values.map { dag ->
      mapOf(
        "data" to
          mapOf(
            "__version" to 3,
            "dag" to serializeDag(dag, fileloc, relativeFileloc),
          ),
      )
    }
  return linkedMapOf(
    "type" to "DagFileParsingResult",
    "fileloc" to fileloc,
    "serialized_dags" to serializedDags,
  )
}

/**
 * Converts a [DagDef] to Airflow DagSerialization v3 format. Required fields are
 * always present; config-driven fields follow the rules in [applyDagConfig]
 * (some always emitted, some only when set).
 */
internal fun serializeDag(
  dag: DagDef,
  fileloc: String,
  relativeFileloc: String,
): Map<String, Any?> {
  val downstream = linkedMapOf<String, MutableList<String>>()
  dag.tasks.forEach { (taskId, def) ->
    def.upstreams.forEach { upstream ->
      downstream.getOrPut(upstream.id) { mutableListOf() } += taskId
    }
  }

  val result =
    linkedMapOf<String, Any?>(
      "dag_id" to dag.id,
      "fileloc" to fileloc,
      "relative_fileloc" to relativeFileloc,
      "timezone" to "UTC",
      "timetable" to serializeTimetable(dag.dagConfig["schedule"] as String?),
      "tasks" to dag.tasks.map { (taskId, def) -> serializeTask(taskId, def, downstream[taskId]) },
      "dag_dependencies" to emptyList<Any?>(),
      "task_group" to serializeTaskGroup(dag.tasks.keys),
      "edge_info" to emptyMap<String, Any?>(),
      "params" to emptyList<Any?>(),
      "deadline" to null,
      "allowed_run_types" to null,
    )
  applyDagConfig(result, dag.dagConfig)
  return result
}

/**
 * Converts one task to the Airflow serialization format. `downstream` is the
 * inverted view of the Dag's upstream edges, sorted for stable JSON.
 *
 * Native Java tasks deliberately emit no `_arg_bindings`: the execution API
 * delivers bindings only for Python `_StubOperator` tasks, and a Java task
 * always executes inside the JVM bundle that already holds its wired inputs,
 * so the runtime resolves them locally.
 */
private fun serializeTask(
  taskId: String,
  def: TaskDef,
  downstream: List<String>?,
): Map<String, Any?> {
  val data =
    linkedMapOf<String, Any?>(
      "task_id" to taskId,
      "task_type" to def.definition.simpleName,
      "_task_module" to def.definition.packageName,
      "language" to "java",
      // Python's operator serializer always emits template_fields (its list
      // value never matches the tuple default it is compared against), so it
      // is unconditional here too. Java tasks have no template fields.
      "template_fields" to emptyList<Any?>(),
    )
  // Emit only config entries that differ from their schema default, mirroring
  // Python BaseSerialization's "omit hard-coded default" behavior. Operator
  // fields are stored unwrapped, so the __type encoding is stripped.
  def.configValues.forEach { (key, value) ->
    if (!matchesSchemaDefault(SchemaFields.TASK[key], value)) {
      data[key] = unwrapTypeEncoding(serializeValue(value))
    }
  }
  if (!downstream.isNullOrEmpty()) {
    data["downstream_task_ids"] = downstream.sorted()
  }
  return mapOf(
    "__type" to "operator",
    "__var" to data,
  )
}

/**
 * Writes Dag-level config onto [data]. Fields with a JSON-schema default
 * (description, dates, tags, fail_fast, ...) are omitted when unset. Fields
 * with no schema default (catchup, disable_bundle_versioning,
 * max_active_tasks, max_active_runs, max_consecutive_failed_dag_runs) are
 * always emitted, because Python's serializer never omits them — it writes
 * the resolved value, falling back to the matching `[core]` config default.
 */
private fun applyDagConfig(
  data: MutableMap<String, Any?>,
  config: Map<String, Any>,
) {
  listOf("description", "dag_display_name", "doc_md", "start_date", "end_date", "dagrun_timeout").forEach { key ->
    config[key]?.let { data[key] = unwrapTypeEncoding(serializeValue(it)) }
  }
  (config["tags"] as? List<*>)?.let { tags ->
    // Python stores tags in a set and serializes them sorted (for a stable
    // dag_hash); mirror that regardless of registration order.
    data["tags"] = tags.map { it.toString() }.sorted()
  }
  data["max_active_tasks"] = config["max_active_tasks"] ?: DEFAULT_MAX_ACTIVE_TASKS_PER_DAG
  data["max_active_runs"] = config["max_active_runs"] ?: DEFAULT_MAX_ACTIVE_RUNS_PER_DAG
  data["max_consecutive_failed_dag_runs"] = config["max_consecutive_failed_dag_runs"] ?: 0
  data["catchup"] = config["catchup"] ?: false
  data["disable_bundle_versioning"] = config["disable_bundle_versioning"] ?: false
  // fail_fast and render_template_as_native_obj have schema default false, so
  // Python omits them when false; keep that behavior.
  if (config["fail_fast"] == true) data["fail_fast"] = true
  if (config["render_template_as_native_obj"] == true) data["render_template_as_native_obj"] = true
  config["is_paused_upon_creation"]?.let { data["is_paused_upon_creation"] = it }
}

// TODO: respect [scheduler] create_cron_data_intervals like Python's
// _create_timetable; the JVM bundle cannot read airflow.cfg, so the
// supervisor must send those flags over the coordinator protocol first.
// Mirrors the Go SDK's default-only behavior; tracked at
// https://github.com/apache/airflow/issues/67938
private fun serializeTimetable(schedule: String?): Map<String, Any?> =
  when (schedule) {
    null -> mapOf("__type" to "airflow.timetables.simple.NullTimetable", "__var" to emptyMap<String, Any?>())
    "@once" -> mapOf("__type" to "airflow.timetables.simple.OnceTimetable", "__var" to emptyMap<String, Any?>())
    "@continuous" ->
      mapOf("__type" to "airflow.timetables.simple.ContinuousTimetable", "__var" to emptyMap<String, Any?>())
    else ->
      mapOf(
        "__type" to "airflow.timetables.trigger.CronTriggerTimetable",
        "__var" to
          mapOf(
            "expression" to schedule,
            "timezone" to "UTC",
            "interval" to 0.0,
            "run_immediately" to false,
          ),
      )
  }

/** Creates the flat root task group containing all task IDs. */
private fun serializeTaskGroup(taskIds: Collection<String>): Map<String, Any?> =
  mapOf(
    "_group_id" to null,
    "group_display_name" to "",
    "prefix_group_id" to true,
    "tooltip" to "",
    "ui_color" to "CornflowerBlue",
    "ui_fgcolor" to "#000",
    "children" to taskIds.associateWith { listOf("operator", it) },
    "upstream_group_ids" to emptyList<Any?>(),
    "downstream_group_ids" to emptyList<Any?>(),
    "upstream_task_ids" to emptyList<Any?>(),
    "downstream_task_ids" to emptyList<Any?>(),
  )

/**
 * Recursively serializes a value with Airflow's type/var encoding, matching
 * Python's `BaseSerialization.serialize()` output: primitives pass through,
 * date-times become `{"__type": "datetime", "__var": epoch_seconds}`,
 * durations `{"__type": "timedelta", "__var": total_seconds}`, and maps
 * `{"__type": "dict", "__var": {...}}`.
 */
internal fun serializeValue(value: Any?): Any? =
  when (value) {
    null -> null
    is String, is Boolean, is Int, is Long, is Double -> value
    is Byte, is Short -> (value as Number).toInt()
    is Float -> value.toDouble()
    is OffsetDateTime -> serializeValue(value.toInstant())
    is Instant ->
      mapOf(
        "__type" to "datetime",
        "__var" to value.epochSecond + value.nano / 1e9,
      )
    is Duration ->
      mapOf(
        "__type" to "timedelta",
        "__var" to value.toNanos() / 1e9,
      )
    is Map<*, *> ->
      mapOf(
        "__type" to "dict",
        "__var" to value.entries.associate { (k, v) -> k.toString() to serializeValue(v) },
      )
    is List<*> -> value.map(::serializeValue)
    is Array<*> -> value.map(::serializeValue)
    else -> value
  }

/**
 * Extracts the `__var` part from a type-encoded value: in Python's
 * `serialize_to_json`, non-decorated fields are serialized then unwrapped.
 */
internal fun unwrapTypeEncoding(value: Any?): Any? {
  val map = value as? Map<*, *> ?: return value
  if ("__type" !in map) return value
  return if ("__var" in map) map["__var"] else value
}

/** Whether a config value equals the schema default and can be omitted. */
private fun matchesSchemaDefault(
  field: Field?,
  value: Any,
): Boolean {
  val defaultJson = field?.defaultJson ?: return false
  val node = defaultsMapper.readTree(defaultJson)
  return when (value) {
    is String -> node.isTextual && node.asText() == value
    is Boolean -> node.isBoolean && node.asBoolean() == value
    is Number -> node.isNumber && node.asDouble() == value.toDouble()
    is Duration -> node.isNumber && node.asDouble() == value.toNanos() / 1e9
    else -> false
  }
}

private fun computeRelativeFileloc(
  fileloc: String,
  bundlePath: String?,
): String {
  if (fileloc.isEmpty()) return ""
  if (bundlePath.isNullOrEmpty()) return "."
  return try {
    Paths
      .get(bundlePath)
      .relativize(Paths.get(fileloc))
      .toString()
      .ifEmpty { "." }
  } catch (e: InvalidPathException) {
    "."
  } catch (e: IllegalArgumentException) {
    "."
  }
}

File: go-sdk/pkg/execution/serde.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package execution

import (
	"fmt"
	"path/filepath"
	"reflect"
	"sort"
	"time"

	"github.com/apache/airflow/go-sdk/bundle/bundlev1"
)

// serializeValue matches Python's type/var encoding.
func serializeValue(value any) any {
	if value == nil {
		return nil
	}
	switch v := value.(type) {
	case string, bool:
		return v
	case int:
		return v
	case int8:
		return int(v)
	case int16:
		return int(v)
	case int32:
		return int(v)
	case int64:
		return v
	case float32:
		return float64(v)
	case float64:
		return v
	case time.Time:
		epochSec := float64(v.Unix()) + float64(v.Nanosecond())/1e9
		return map[string]any{
			"__type": "datetime",
			"__var":  epochSec,
		}
	case time.Duration:
		return map[string]any{
			"__type": "timedelta",
			"__var":  v.Seconds(),
		}
	case map[string]any:
		serialized := make(map[string]any, len(v))
		for k, val := range v {
			serialized[k] = serializeValue(val)
		}
		return map[string]any{
			"__type": "dict",
			"__var":  serialized,
		}
	case []string:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = serializeValue(item)
		}
		return result
	case []any:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = serializeValue(item)
		}
		return result
	default:
		rv := reflect.ValueOf(v)
		switch rv.Kind() {
		case reflect.Map:
			serialized := make(map[string]any, rv.Len())
			for _, key := range rv.MapKeys() {
				serialized[fmt.Sprint(key.Interface())] = serializeValue(rv.MapIndex(key).Interface())
			}
			return map[string]any{
				"__type": "dict",
				"__var":  serialized,
			}
		case reflect.Slice, reflect.Array:
			result := make([]any, rv.Len())
			for i := range result {
				result[i] = serializeValue(rv.Index(i).Interface())
			}
			return result
		default:
			return v
		}
	}
}

// unwrapTypeEncoding strips type/var metadata.
func unwrapTypeEncoding(value any) any {
	m, ok := value.(map[string]any)
	if !ok {
		return value
	}
	if _, hasType := m["__type"]; !hasType {
		return value
	}
	if v, hasVar := m["__var"]; hasVar {
		return v
	}
	return value
}

// serializeTimetable converts a schedule string to the Airflow timetable format.
// TODO: honor scheduler interval flags once the supervisor sends them.
// Tracked at https://github.com/apache/airflow/issues/67938
func serializeTimetable(schedule *string) map[string]any {
	if schedule == nil {
		return map[string]any{
			"__type": "airflow.timetables.simple.NullTimetable",
			"__var":  map[string]any{},
		}
	}
	switch *schedule {
	case "@once":
		return map[string]any{
			"__type": "airflow.timetables.simple.OnceTimetable",
			"__var":  map[string]any{},
		}
	case "@continuous":
		return map[string]any{
			"__type": "airflow.timetables.simple.ContinuousTimetable",
			"__var":  map[string]any{},
		}
	default:
		return map[string]any{
			"__type": "airflow.timetables.trigger.CronTriggerTimetable",
			"__var": map[string]any{
				"expression":      *schedule,
				"timezone":        "UTC",
				"interval":        0.0,
				"run_immediately": false,
			},
		}
	}
}

// serializeTask builds a serialized task.
func serializeTask(info bundlev1.TaskInfo) map[string]any {
	typeName := info.TypeName
	if typeName == "" {
		typeName = info.ID
	}
	pkgPath := info.PkgPath
	if pkgPath == "" {
		pkgPath = "main"
	}
	data := map[string]any{
		"task_id":      info.ID,
		"task_type":    typeName,
		"_task_module": pkgPath,
		"language":     "go",
		// Python always emits this; Go tasks have no template fields.
		"template_fields": []any{},
	}
	for key, value := range info.Spec.SchemaFields() {
		data[key] = unwrapTypeEncoding(serializeValue(value))
	}
	if len(info.Downstream) > 0 {
		sorted := make([]string, len(info.Downstream))
		copy(sorted, info.Downstream)
		sort.Strings(sorted)
		data["downstream_task_ids"] = sorted
	}
	return map[string]any{
		"__type": "operator",
		"__var":  data,
	}
}

// applyDagSpec writes serialized Dag fields.
func applyDagSpec(data map[string]any, s bundlev1.DagSpec) {
	for key, value := range s.SchemaFields() {
		data[key] = unwrapTypeEncoding(serializeValue(value))
	}
}

// serializeTaskGroup creates a flat root task group containing all task IDs.
func serializeTaskGroup(taskIDs []string) map[string]any {
	children := make(map[string]any, len(taskIDs))
	for _, id := range taskIDs {
		children[id] = []any{"operator", id}
	}
	return map[string]any{
		"_group_id":            nil,
		"group_display_name":   "",
		"prefix_group_id":      true,
		"tooltip":              "",
		"ui_color":             "CornflowerBlue",
		"ui_fgcolor":           "#000",
		"children":             children,
		"upstream_group_ids":   []any{},
		"downstream_group_ids": []any{},
		"upstream_task_ids":    []any{},
		"downstream_task_ids":  []any{},
	}
}

// serializeParams builds serialized Dag params.
func serializeParams(params map[string]any) []any {
	if len(params) == 0 {
		return []any{}
	}
	result := make([]any, 0, len(params))
	for k, v := range params {
		result = append(result, []any{
			k,
			map[string]any{
				"__class":     "airflow.sdk.definitions.param.Param",
				"default":     serializeValue(v),
				"description": nil,
				"schema":      serializeValue(map[string]any{}),
				"source":      nil,
			},
		})
	}
	return result
}

// SerializeDag converts DagInfo to DagSerialization v3.
func SerializeDag(info bundlev1.DagInfo, fileloc, relativeFileloc string) map[string]any {
	taskIDs := make([]string, len(info.Tasks))
	tasks := make([]any, len(info.Tasks))
	for i, t := range info.Tasks {
		taskIDs[i] = t.ID
		tasks[i] = serializeTask(t)
	}

	var schedule *string
	if info.Spec.Schedule != "" {
		s := info.Spec.Schedule
		schedule = &s
	}

	result := map[string]any{
		"dag_id":            info.DagID,
		"fileloc":           fileloc,
		"relative_fileloc":  relativeFileloc,
		"timezone":          "UTC",
		"timetable":         serializeTimetable(schedule),
		"tasks":             tasks,
		"dag_dependencies":  []any{},
		"task_group":        serializeTaskGroup(taskIDs),
		"edge_info":         map[string]any{},
		"params":            serializeParams(nil),
		"deadline":          nil,
		"allowed_run_types": nil,
	}
	applyDagSpec(result, info.Spec)
	return result
}

// computeRelativeFileloc returns a path relative to the bundle.
func computeRelativeFileloc(fileloc, bundlePath string) string {
	if fileloc == "" {
		return ""
	}
	if bundlePath == "" {
		return "."
	}
	rel, err := filepath.Rel(bundlePath, fileloc)
	if err != nil {
		return "."
	}
	if rel == "" {
		return "."
	}
	return rel
}
